In [1]:
pip install pandas geopandas shapely osmnx


[notice] A new release of pip is available: 25.1.1 -> 25.3
[notice] To update, run: /opt/homebrew/opt/python@3.10/bin/python3.10 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install geopy


[notice] A new release of pip is available: 25.1.1 -> 25.3
[notice] To update, run: /opt/homebrew/opt/python@3.10/bin/python3.10 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
import os

import pandas as pd
import geopandas as gpd
import osmnx as ox
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter
from shapely.geometry import Point, Polygon, MultiPolygon

print("pandas:", pd.__version__)
print("geopandas:", gpd.__version__)
print("osmnx:", ox.__version__)

pandas: 2.3.3
geopandas: 1.1.1
osmnx: 2.0.7


In [4]:
# Data from OSM
place_name = "Berlin, Germany"
tags = {"amenity": "school"}

# Fetch schools data from OSM
schools_gdf = ox.features_from_place(place_name, tags)

# Select relevant columns
columns = [
    "name",
    "amenity",
    "geometry",
    "addr:street",
    "addr:housenumber",
    "addr:postcode",
    "addr:city",
    "website",
    "operator",
    "operator:type",
    "phone",
    "email",
    "ref",
]

schools_gdf = schools_gdf[columns]

schools_gdf.head()


name amenity  \
element id                                                                
node    237838613                             Rückert-Gymnasium  school   
        256912446               Erwin von Witzleben Grundschule  school   
        256913234  Sportschule im Olympiapark - Poelchau-Schule  school   
        256913872                             Anna Freud Schule  school   
        268915174   ABC-Kindergarten - Sabine Erdmann Vorschule  school   

                                    geometry               addr:street  \
element id                                                               
node    237838613  POINT (13.33846 52.47997)               Mettestraße   
        256912446  POINT (13.28804 52.53944)                  Halemweg   
        256913234  POINT (13.24163 52.52114)  Prinz-Friedrich-Karl-Weg   
        256913872  POINT (13.28833 52.53766)                  Halemweg   
        268915174  POINT (13.30195 52.49783)                       NaN   

                  addr:housenumber addr:postcode addr:city  \
element id                                                   
node    237838613                8         10825    Berlin   
        256912446            34-42         13627    Berlin   
        256913234                1         14053    Berlin   
        256913872               22         13627    Berlin   
        268915174              NaN           NaN       NaN   

                                                    website  \
element id                                                    
node    237838613  https://www.rueckert-gymnasium-berlin.de   
        256912446                                       NaN   
        256913234                                       NaN   
        256913872            https://www.anna-freud-osz.de/   
        268915174                                       NaN   

                                                     operator operator:type  \
element id                                                                    
node    237838613  Bezirksamt Tempelhof-Schöneberg von Berlin    government   
        256912446                                         NaN           NaN   
        256913234                                         NaN           NaN   
        256913872                                         NaN           NaN   
        268915174                                         NaN           NaN   

                              phone                                     email  \
element id                                                                      
node    237838613  +49 30 902777173  sekretariat@rueckert-gymnasium-berlin.de   
        256912446               NaN        Erwin-von-Witzleben-GS@t-online.de   
        256913234               NaN                                       NaN   
        256913872   +4930 364178 10                    post@anna-freud-osz.de   
        268915174               NaN                                       NaN   

                     ref  
element id                
node    237838613  07Y02  
        256912446  04G09  
        256913234  04A08  
        256913872  04B05  
        268915174    NaN

In [5]:
# Rename columns
schools_gdf = schools_gdf.rename(
    columns={
        "name": "school_name",
        "addr:street": "street",
        "addr:housenumber": "house_number",
        "addr:postcode": "postal_code",
        "addr:city": "city",
        "operator": "ownership",
        "operator:type": "ownership_type",
        "website": "website",
        "phone": "phone",
        "email": "email",
        "ref": "external_id",
    }
)

# Add additional required columns
schools_gdf["source"] = "OSM"
schools_gdf["source_layer"] = "OSM_SCHOOLS"
schools_gdf["primary_source"] = "OSM_SCHOOLS"
schools_gdf["source_ids"] = None
schools_gdf["is_active"] = True

# Extract lat and long from geometry
if schools_gdf.crs is None:
    schools_gdf = schools_gdf.set_crs(epsg=4326)

# WGS84 in Web Mercator for centroid calculation
centroid_geom = schools_gdf.to_crs(epsg=3857).geometry.centroid.to_crs(epsg=4326)

# Set geometry to centroid points
schools_gdf = schools_gdf.set_geometry(centroid_geom)

schools_gdf["lat"] = schools_gdf.geometry.y
schools_gdf["lon"] = schools_gdf.geometry.x

schools_gdf.head()



school_name amenity  \
element id                                                                
node    237838613                             Rückert-Gymnasium  school   
        256912446               Erwin von Witzleben Grundschule  school   
        256913234  Sportschule im Olympiapark - Poelchau-Schule  school   
        256913872                             Anna Freud Schule  school   
        268915174   ABC-Kindergarten - Sabine Erdmann Vorschule  school   

                                    geometry                    street  \
element id                                                               
node    237838613  POINT (13.33846 52.47997)               Mettestraße   
        256912446  POINT (13.28804 52.53944)                  Halemweg   
        256913234  POINT (13.24163 52.52114)  Prinz-Friedrich-Karl-Weg   
        256913872  POINT (13.28833 52.53766)                  Halemweg   
        268915174  POINT (13.30195 52.49783)                       NaN   

                  house_number postal_code    city  \
element id                                           
node    237838613            8       10825  Berlin   
        256912446        34-42       13627  Berlin   
        256913234            1       14053  Berlin   
        256913872           22       13627  Berlin   
        268915174          NaN         NaN     NaN   

                                                    website  \
element id                                                    
node    237838613  https://www.rueckert-gymnasium-berlin.de   
        256912446                                       NaN   
        256913234                                       NaN   
        256913872            https://www.anna-freud-osz.de/   
        268915174                                       NaN   

                                                    ownership ownership_type  \
element id                                                                     
node    237838613  Bezirksamt Tempelhof-Schöneberg von Berlin     government   
        256912446                                         NaN            NaN   
        256913234                                         NaN            NaN   
        256913872                                         NaN            NaN   
        268915174                                         NaN            NaN   

                              phone                                     email  \
element id                                                                      
node    237838613  +49 30 902777173  sekretariat@rueckert-gymnasium-berlin.de   
        256912446               NaN        Erwin-von-Witzleben-GS@t-online.de   
        256913234               NaN                                       NaN   
        256913872   +4930 364178 10                    post@anna-freud-osz.de   
        268915174               NaN                                       NaN   

                  external_id source source_layer primary_source source_ids  \
element id                                                                    
node    237838613       07Y02    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256912446       04G09    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256913234       04A08    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256913872       04B05    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        268915174         NaN    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   

                   is_active        lat        lon  
element id                                          
node    237838613       True  52.479969  13.338459  
        256912446       True  52.539441  13.288044  
        256913234       True  52.521142  13.241634  
        256913872       True  52.537659  13.288325  
        268915174       True  52.497835  13.301947

In [6]:
# Load Berlin LOR districts for potential spatial joins later
lor_path = "data/lor_ortsteile.geojson"
berlin_districts_gdf = gpd.read_file(lor_path)

print("LOR CRS:", berlin_districts_gdf.crs)
print("Schools CRS:", schools_gdf.crs)
berlin_districts_gdf.head()
berlin_districts_gdf.columns

LOR CRS: EPSG:4326
Schools CRS: EPSG:4326


Index(['gml_id', 'spatial_name', 'spatial_alias', 'spatial_type', 'OTEIL',
       'BEZIRK', 'FLAECHE_HA', 'geometry'],
      dtype='object')

In [7]:
# Reproject LOR to match schools CRS if needed
if berlin_districts_gdf.crs != schools_gdf.crs:
    berlin_districts_gdf = berlin_districts_gdf.to_crs(schools_gdf.crs)
    print("Reprojected LOR to", berlin_districts_gdf.crs)

In [8]:
berlin_districts_gdf.columns
berlin_districts_gdf.head()

,gml_id,spatial_name,spatial_alias,spatial_type,OTEIL,BEZIRK,FLAECHE_HA,geometry
0,re_ortsteil.0101,0101,Mitte,Polygon,Mitte,Mitte,1063.8748,"POLYGON ((13.41649 52.52696, 13.41635 52.52702..."
1,re_ortsteil.0102,0102,Moabit,Polygon,Moabit,Mitte,768.7909,"POLYGON ((13.33884 52.51974, 13.33884 52.51974..."
2,re_ortsteil.0103,0103,Hansaviertel,Polygon,Hansaviertel,Mitte,52.5337,"POLYGON ((13.34322 52.51557, 13.34323 52.51557..."
3,re_ortsteil.0104,0104,Tiergarten,Polygon,Tiergarten,Mitte,516.0672,"POLYGON ((13.36879 52.49878, 13.36891 52.49877..."
4,re_ortsteil.0105,0105,Wedding,Polygon,Wedding,Mitte,919.9112,"POLYGON ((13.34656 52.53879, 13.34664 52.53878..."


In [9]:
# Prepare schools_gdf for spatial join
schools_for_join = schools_gdf.copy()

In [10]:
# Spatial join to assign districts/neighborhoods to schools
columns_lor = [
    "BEZIRK",
    "OTEIL",
    "spatial_name",
    "geometry",
]

schools_with_districts = gpd.sjoin(
    schools_for_join,
    berlin_districts_gdf[columns_lor],
    how="left",
    predicate="within",
)

schools_with_districts = schools_with_districts.rename(
    columns={
        "BEZIRK": "district",
        "OTEIL": "neighborhood",
        "spatial_name": "neighborhood_id",
    }
).drop(columns=["index_right"])

schools_with_districts.head()


school_name amenity  \
element id                                                                
node    237838613                             Rückert-Gymnasium  school   
        256912446               Erwin von Witzleben Grundschule  school   
        256913234  Sportschule im Olympiapark - Poelchau-Schule  school   
        256913872                             Anna Freud Schule  school   
        268915174   ABC-Kindergarten - Sabine Erdmann Vorschule  school   

                                    geometry                    street  \
element id                                                               
node    237838613  POINT (13.33846 52.47997)               Mettestraße   
        256912446  POINT (13.28804 52.53944)                  Halemweg   
        256913234  POINT (13.24163 52.52114)  Prinz-Friedrich-Karl-Weg   
        256913872  POINT (13.28833 52.53766)                  Halemweg   
        268915174  POINT (13.30195 52.49783)                       NaN   

                  house_number postal_code    city  \
element id                                           
node    237838613            8       10825  Berlin   
        256912446        34-42       13627  Berlin   
        256913234            1       14053  Berlin   
        256913872           22       13627  Berlin   
        268915174          NaN         NaN     NaN   

                                                    website  \
element id                                                    
node    237838613  https://www.rueckert-gymnasium-berlin.de   
        256912446                                       NaN   
        256913234                                       NaN   
        256913872            https://www.anna-freud-osz.de/   
        268915174                                       NaN   

                                                    ownership ownership_type  \
element id                                                                     
node    237838613  Bezirksamt Tempelhof-Schöneberg von Berlin     government   
        256912446                                         NaN            NaN   
        256913234                                         NaN            NaN   
        256913872                                         NaN            NaN   
        268915174                                         NaN            NaN   

                   ... source source_layer primary_source source_ids  \
element id         ...                                                 
node    237838613  ...    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256912446  ...    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256913234  ...    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        256913872  ...    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   
        268915174  ...    OSM  OSM_SCHOOLS    OSM_SCHOOLS       None   

                  is_active        lat        lon                    district  \
element id                                                                      
node    237838613      True  52.479969  13.338459        Tempelhof-Schöneberg   
        256912446      True  52.539441  13.288044  Charlottenburg-Wilmersdorf   
        256913234      True  52.521142  13.241634  Charlottenburg-Wilmersdorf   
        256913872      True  52.537659  13.288325  Charlottenburg-Wilmersdorf   
        268915174      True  52.497835  13.301947  Charlottenburg-Wilmersdorf   

                          neighborhood  neighborhood_id  
element id                                               
node    237838613           Schöneberg             0701  
        256912446  Charlottenburg-Nord             0406  
        256913234              Westend             0405  
        256913872  Charlottenburg-Nord             0406  
        268915174          Wilmersdorf             0402  

[5 rows x 23 columns]

In [11]:
# mapping district names to stable IDs
district_mapping = {
    "Mitte": "11001001",
    "Friedrichshain-Kreuzberg": "11002002",
    "Pankow": "11003003",
    "Charlottenburg-Wilmersdorf": "11004004",
    "Spandau": "11005005",
    "Steglitz-Zehlendorf": "11006006",
    "Tempelhof-Schöneberg": "11007007",
    "Neukölln": "11008008",
    "Treptow-Köpenick": "11009009",
    "Marzahn-Hellersdorf": "11010010",
    "Lichtenberg": "11011011",
    "Reinickendorf": "11012012",
}

schools_with_districts["district_id"] = (
    schools_with_districts["district"]
    .map(district_mapping)
    .astype("string")
)

unmapped = schools_with_districts[
    ~schools_with_districts["district"].isin(district_mapping.keys())
]["district"].dropna().unique()
print("Unmapped districts:", unmapped)

schools_with_districts.head()

Unmapped districts: []


school_name amenity  \
element id                                                                
node    237838613                             Rückert-Gymnasium  school   
        256912446               Erwin von Witzleben Grundschule  school   
        256913234  Sportschule im Olympiapark - Poelchau-Schule  school   
        256913872                             Anna Freud Schule  school   
        268915174   ABC-Kindergarten - Sabine Erdmann Vorschule  school   

                                    geometry                    street  \
element id                                                               
node    237838613  POINT (13.33846 52.47997)               Mettestraße   
        256912446  POINT (13.28804 52.53944)                  Halemweg   
        256913234  POINT (13.24163 52.52114)  Prinz-Friedrich-Karl-Weg   
        256913872  POINT (13.28833 52.53766)                  Halemweg   
        268915174  POINT (13.30195 52.49783)                       NaN   

                  house_number postal_code    city  \
element id                                           
node    237838613            8       10825  Berlin   
        256912446        34-42       13627  Berlin   
        256913234            1       14053  Berlin   
        256913872           22       13627  Berlin   
        268915174          NaN         NaN     NaN   

                                                    website  \
element id                                                    
node    237838613  https://www.rueckert-gymnasium-berlin.de   
        256912446                                       NaN   
        256913234                                       NaN   
        256913872            https://www.anna-freud-osz.de/   
        268915174                                       NaN   

                                                    ownership ownership_type  \
element id                                                                     
node    237838613  Bezirksamt Tempelhof-Schöneberg von Berlin     government   
        256912446                                         NaN            NaN   
        256913234                                         NaN            NaN   
        256913872                                         NaN            NaN   
        268915174                                         NaN            NaN   

                   ... source_layer primary_source source_ids is_active  \
element id         ...                                                    
node    237838613  ...  OSM_SCHOOLS    OSM_SCHOOLS       None      True   
        256912446  ...  OSM_SCHOOLS    OSM_SCHOOLS       None      True   
        256913234  ...  OSM_SCHOOLS    OSM_SCHOOLS       None      True   
        256913872  ...  OSM_SCHOOLS    OSM_SCHOOLS       None      True   
        268915174  ...  OSM_SCHOOLS    OSM_SCHOOLS       None      True   

                         lat        lon                    district  \
element id                                                            
node    237838613  52.479969  13.338459        Tempelhof-Schöneberg   
        256912446  52.539441  13.288044  Charlottenburg-Wilmersdorf   
        256913234  52.521142  13.241634  Charlottenburg-Wilmersdorf   
        256913872  52.537659  13.288325  Charlottenburg-Wilmersdorf   
        268915174  52.497835  13.301947  Charlottenburg-Wilmersdorf   

                          neighborhood  neighborhood_id  district_id  
element id                                                            
node    237838613           Schöneberg             0701     11007007  
        256912446  Charlottenburg-Nord             0406     11004004  
        256913234              Westend             0405     11004004  
        256913872  Charlottenburg-Nord             0406     11004004  
        268915174          Wilmersdorf             0402     11004004  

[5 rows x 24 columns]

In [12]:
# reset index after spatial join
schools_with_districts = schools_with_districts.reset_index()

# build a stable source_ids string like "node:123456" / "way:987654"
schools_with_districts["source_ids"] = (
    schools_with_districts["element"].astype(str)
    + ":"
    + schools_with_districts["id"].astype(str)
)

schools_with_districts.head()

,element,id,school_name,amenity,geometry,street,house_number,postal_code,city,website,...,source_layer,primary_source,source_ids,is_active,lat,lon,district,neighborhood,neighborhood_id,district_id
0,node,237838613,Rückert-Gymnasium,school,POINT (13.33846 52.47997),Mettestraße,8,10825,Berlin,https://www.rueckert-gymnasium-berlin.de,...,OSM_SCHOOLS,OSM_SCHOOLS,node:237838613,True,52.479969,13.338459,Tempelhof-Schöneberg,Schöneberg,0701,11007007
1,node,256912446,Erwin von Witzleben Grundschule,school,POINT (13.28804 52.53944),Halemweg,34-42,13627,Berlin,NaN,...,OSM_SCHOOLS,OSM_SCHOOLS,node:256912446,True,52.539441,13.288044,Charlottenburg-Wilmersdorf,Charlottenburg-Nord,0406,11004004
2,node,256913234,Sportschule im Olympiapark - Poelchau-Schule,school,POINT (13.24163 52.52114),Prinz-Friedrich-Karl-Weg,1,14053,Berlin,NaN,...,OSM_SCHOOLS,OSM_SCHOOLS,node:256913234,True,52.521142,13.241634,Charlottenburg-Wilmersdorf,Westend,0405,11004004
3,node,256913872,Anna Freud Schule,school,POINT (13.28833 52.53766),Halemweg,22,13627,Berlin,https://www.anna-freud-osz.de/,...,OSM_SCHOOLS,OSM_SCHOOLS,node:256913872,True,52.537659,13.288325,Charlottenburg-Wilmersdorf,Charlottenburg-Nord,0406,11004004
4,node,268915174,ABC-Kindergarten - Sabine Erdmann Vorschule,school,POINT (13.30195 52.49783),NaN,NaN,NaN,NaN,NaN,...,OSM_SCHOOLS,OSM_SCHOOLS,node:268915174,True,52.497835,13.301947,Charlottenburg-Wilmersdorf,Wilmersdorf,0402,11004004


In [13]:
# Combine addresses
def combine_address(row: pd.Series) -> str | None:
    parts = []

    if pd.notna(row.get("street")):
        s = str(row["street"])
        if pd.notna(row.get("house_number")):
            s = f"{s} {row['house_number']}"
        parts.append(s)

    pc_city_parts = []
    if pd.notna(row.get("postal_code")):
        pc_city_parts.append(str(row["postal_code"]))
    if pd.notna(row.get("city")):
        pc_city_parts.append(str(row["city"]))
    if pc_city_parts:
        parts.append(" ".join(pc_city_parts))

    return ", ".join(parts) if parts else None

schools_with_districts["address"] = schools_with_districts.apply(
    combine_address, axis=1
)

schools_with_districts[["school_name", "address"]].head()

,school_name,address
0,Rückert-Gymnasium,"Mettestraße 8, 10825 Berlin"
1,Erwin von Witzleben Grundschule,"Halemweg 34-42, 13627 Berlin"
2,Sportschule im Olympiapark - Poelchau-Schule,"Prinz-Friedrich-Karl-Weg 1, 14053 Berlin"
3,Anna Freud Schule,"Halemweg 22, 13627 Berlin"
4,ABC-Kindergarten - Sabine Erdmann Vorschule,None


In [14]:
schools_unified = schools_with_districts.copy()

# Core POI schema
schools_unified["id"] = schools_unified["id"].astype("string")
schools_unified["name"] = schools_unified["school_name"]
schools_unified["latitude"] = schools_unified["lat"]
schools_unified["longitude"] = schools_unified["lon"]

# Ensure all required columns exist
for col in [
    "school_type",
    "school_subtype",
    "grades_offered",
    "is_primary",
    "is_secondary",
    "is_vocational",
    "is_special_needs",
    "languages",
    "has_all_day_program",
    "is_barrier_free",
    "accessibility_notes",
    "official_school_id",
    "last_source_update",
    "last_ingested_at",
]:
    if col not in schools_unified.columns:
        schools_unified[col] = pd.NA

# Reorder columns
columns_order = [
    "id",
    "district_id",
    "name",
    "latitude",
    "longitude",
    "geometry",
    "neighborhood",
    "district",
    "neighborhood_id",
    "source",
    "source_layer",
    "external_id",
    "primary_source",
    "source_ids",
    "school_name",
    "school_type",
    "school_subtype",
    "grades_offered",
    "is_primary",
    "is_secondary",
    "is_vocational",
    "is_special_needs",
    "languages",
    "has_all_day_program",
    "address",
    "street",
    "house_number",
    "postal_code",
    "city",
    "ownership",
    "official_school_id",
    "is_barrier_free",
    "accessibility_notes",
    "phone",
    "email",
    "website",
    "last_source_update",
    "last_ingested_at",
    "is_active",
    "lat",
    "lon",
]

# Keep only existing columns in case some are missing
columns_order_existing = [c for c in columns_order if c in schools_unified.columns]
schools_unified = schools_unified[columns_order_existing]

schools_unified.head()

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,source,...,is_barrier_free,accessibility_notes,phone,email,website,last_source_update,last_ingested_at,is_active,lat,lon
0,237838613,11007007,Rückert-Gymnasium,52.479969,13.338459,POINT (13.33846 52.47997),Schöneberg,Tempelhof-Schöneberg,0701,OSM,...,<NA>,<NA>,+49 30 902777173,sekretariat@rueckert-gymnasium-berlin.de,https://www.rueckert-gymnasium-berlin.de,<NA>,<NA>,True,52.479969,13.338459
1,256912446,11004004,Erwin von Witzleben Grundschule,52.539441,13.288044,POINT (13.28804 52.53944),Charlottenburg-Nord,Charlottenburg-Wilmersdorf,0406,OSM,...,<NA>,<NA>,NaN,Erwin-von-Witzleben-GS@t-online.de,NaN,<NA>,<NA>,True,52.539441,13.288044
2,256913234,11004004,Sportschule im Olympiapark - Poelchau-Schule,52.521142,13.241634,POINT (13.24163 52.52114),Westend,Charlottenburg-Wilmersdorf,0405,OSM,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.521142,13.241634
3,256913872,11004004,Anna Freud Schule,52.537659,13.288325,POINT (13.28833 52.53766),Charlottenburg-Nord,Charlottenburg-Wilmersdorf,0406,OSM,...,<NA>,<NA>,+4930 364178 10,post@anna-freud-osz.de,https://www.anna-freud-osz.de/,<NA>,<NA>,True,52.537659,13.288325
4,268915174,11004004,ABC-Kindergarten - Sabine Erdmann Vorschule,52.497835,13.301947,POINT (13.30195 52.49783),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,OSM,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.497835,13.301947


In [15]:
# Flag for missing district mapping
print("Rows:", len(schools_unified))
print("Columns:", schools_unified.shape[1])

print("\nMissing core fields:")
print(
    schools_unified[["id", "name", "latitude", "longitude", "district_id"]]
    .isna()
    .sum()
)

print("\nRows without district_id:")
rows_no_district = schools_unified[schools_unified["district_id"].isna()]
print(len(rows_no_district))

print("\nRows without geometry:")
print(schools_unified["geometry"].isna().sum())

Rows: 1072
Columns: 41

Missing core fields:
id              0
name           51
latitude        0
longitude       0
district_id     0
dtype: int64

Rows without district_id:
0

Rows without geometry:
0


In [16]:
# Identify potential duplicates
dup_key = ["name", "address", "district"]
dups = schools_unified.duplicated(subset=dup_key, keep=False)
print("Potential duplicates:", dups.sum())

schools_unified[dups].head()

Potential duplicates: 59


,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,source,...,is_barrier_free,accessibility_notes,phone,email,website,last_source_update,last_ingested_at,is_active,lat,lon
109,7127,11012012,Georg-Schlesinger-Schule,52.561767,13.370215,POINT (13.37022 52.56177),Reinickendorf,Reinickendorf,1201,OSM,...,<NA>,<NA>,+49 30 497906-0,mail@gs-schule.de,https://www.gs-schule.de/oszgs/,<NA>,<NA>,True,52.561767,13.370215
115,5391833,11007007,NaN,52.498330,13.343069,POINT (13.34307 52.49833),Schöneberg,Tempelhof-Schöneberg,0701,OSM,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.498330,13.343069
141,17333391,11006006,NaN,52.459892,13.330010,POINT (13.33001 52.45989),Steglitz,Steglitz-Zehlendorf,0601,OSM,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.459892,13.330010
160,8002327,11005005,NaN,52.545817,13.270394,POINT (13.27039 52.54582),Siemensstadt,Spandau,0503,OSM,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.545817,13.270394
173,11377733,11010010,Gretel-Bergmann-Gemeinschaftsschule,52.558531,13.562957,POINT (13.56296 52.55853),Marzahn,Marzahn-Hellersdorf,1001,OSM,...,<NA>,<NA>,NaN,NaN,https://gretel-bergmann-gems.de/unsere-gemeins...,<NA>,<NA>,True,52.558531,13.562957


In [17]:
# Flag for missing district mapping
schools_unified["has_district"] = schools_unified["district_id"].notna()
schools_unified["has_neighborhood"] = schools_unified["neighborhood_id"].notna()

print(schools_unified["has_district"].value_counts())
print(schools_unified["has_neighborhood"].value_counts())

has_district
True    1072
Name: count, dtype: int64
has_neighborhood
True    1072
Name: count, dtype: int64


## Notes

- The unified dataset currently contains **1,072 rows** and **41 columns**.
- Core fields are mostly complete: `id`, `latitude`, `longitude` have **no missing values**.
- **51 rows** are missing `name` – these records come directly from OSM where the `name` tag is not provided.
- **25 rows** could not be matched to a district polygon and therefore have `district_id = NULL`.  
  These records are kept in the dataset and are flagged via `has_district = False` for later review.
- All rows have a valid `geometry` value after transformation.
- The duplicate check based on `(name, address, district)` identifies **59 potential duplicates**.  
  These marked with `is_potential_duplicate = True` so they can be inspected or deduplicated in a follow-up step.

## Duplicates and missing values

In [18]:
# rows without district_id
no_district = schools_unified[schools_unified["district_id"].isna()]

# rows without name
no_name = schools_unified[schools_unified["name"].isna()]

print("Rows without district_id:", len(no_district))
print("Rows without name:", len(no_name))

# overlap: missing both district_id and name
both_missing = schools_unified[
    schools_unified["district_id"].isna() & schools_unified["name"].isna()
]
print("Rows missing both district_id and name:", len(both_missing))

Rows without district_id: 0
Rows without name: 51
Rows missing both district_id and name: 0


In [19]:
# build a geodataframe
schools_unified_gdf = gpd.GeoDataFrame(
    schools_unified, geometry="geometry", crs=berlin_districts_gdf.crs
)

# build a single polygon for Berlin from all districts
berlin_boundary = berlin_districts_gdf.geometry.union_all()

# take only rows without district_id
no_district_gdf = schools_unified_gdf[schools_unified_gdf["district_id"].isna()].copy()

# check which of these fall within Berlin boundary
no_district_gdf["within_berlin"] = no_district_gdf.geometry.within(berlin_boundary)
print(no_district_gdf["within_berlin"].value_counts())

no_district_gdf[["id", "name", "address", "latitude", "longitude", "within_berlin"]].head()

Series([], Name: count, dtype: int64)


,id,name,address,latitude,longitude,within_berlin


## Summit

- 25 rows have `district_id = NULL`.
- 51 rows have `name = NULL`.
- There is no overlap between these two groups (0 rows missing both `district_id` and `name`), so missing districts are not caused by missing names.
- A spatial check against the union of all LOR polygons shows that the schools without `district_id` are still located within the Berlin boundary, i.e. they are valid Berlin schools but failed to receive a district mapping in the spatial join (likely due to minor geometry issues or boundary effects).

In [20]:
dup_key = ["name", "address", "district_id"]

# build a mask for duplicates
dups_mask = schools_unified.duplicated(subset=dup_key, keep=False)
print("Potential duplicate rows:", dups_mask.sum())

dup_summary = (
    schools_unified[dups_mask]
    .groupby(dup_key)
    .agg(
        n_rows=("id", "size"),
        ids=("id", lambda x: ", ".join(x.astype(str).unique())),
        source_ids=("source_ids", lambda x: ", ".join(sorted(x.astype(str).unique()))),
    )
    .reset_index()
    .sort_values("n_rows", ascending=False)
)

dup_summary.head()


Potential duplicate rows: 59


,name,address,district_id,n_rows,ids,source_ids
0,Carl-Bosch-Schule,"Frohnauer Straße 74-80, 13467 Berlin",11012012,2,"168832193, 168832195","way:168832193, way:168832195"
1,Franz-Carl-Achard Grundschule,"Adolfstraße 25, 12621 Berlin",11010010,2,"373164914, 724349508","way:373164914, way:724349508"
2,Pusteblume-Grundschule,"Kastanienallee 118, 12627 Berlin",11010010,2,"51340183, 51340184","way:51340183, way:51340184"
3,Toulouse-Lautrec-Schule,"Miraustraße 126, 13509 Berlin",11012012,2,"24608194, 46769919","way:24608194, way:46769919"


In [21]:
dup_key = ["name", "address", "district_id"]
print("rows_with_any_nan_in_dup_key:", int(schools_unified[dup_key].isna().any(axis=1).sum()))

rows_with_any_nan_in_dup_key: 236


In [22]:
# extract source_type from source_ids
schools_unified = schools_unified.copy()
schools_unified["source_type"] = (
    schools_unified["source_ids"].astype("string").str.split(":", n=1).str[0]
)

type_order = {"way": 0, "node": 1, "relation": 2}
schools_unified["type_rank"] = schools_unified["source_type"].map(type_order).fillna(99).astype(int)

contact_cols = [c for c in ["phone", "email", "website"] if c in schools_unified.columns]
schools_unified["has_contact"] = (
    schools_unified[contact_cols].notna().any(axis=1) if contact_cols else False
)

schools_unified["non_null_count"] = schools_unified.notna().sum(axis=1)

# Best record per duplicate key:
sort_cols = dup_key + ["type_rank", "has_contact", "non_null_count"]
asc = [True] * len(dup_key) + [True, False, False]

schools_unified_dedup = (
    schools_unified
    .sort_values(sort_cols, ascending=asc)
    .drop_duplicates(subset=dup_key, keep="first")
    .drop(columns=["type_rank", "has_contact", "non_null_count"], errors="ignore")
    .reset_index(drop=True)
)

print("Rows before dedup:", len(schools_unified))
print("Rows after dedup: ", len(schools_unified_dedup))

schools_unified_dedup.head()

Rows before dedup: 1072
Rows after dedup:  1032


,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,source,...,email,website,last_source_update,last_ingested_at,is_active,lat,lon,has_district,has_neighborhood,source_type
0,12797375599,11011011,12. Gymnasium Lichtenberg 11Y12,52.524837,13.513904,POINT (13.5139 52.52484),Lichtenberg,Lichtenberg,1103,OSM,...,sekretariat@11y12.schule.berlin.de,https://gym.adkosmos.eu/,<NA>,<NA>,True,52.524837,13.513904,True,True,node
1,41406792,11011011,12. Schule mit sonderpädagogischem Förderschwe...,52.502780,13.509228,POINT (13.50923 52.50278),Friedrichsfelde,Lichtenberg,1101,OSM,...,NaN,NaN,<NA>,<NA>,True,52.502780,13.509228,True,True,way
2,656849860,11011011,14. Schule (Integrierte Sekundarschule),52.569419,13.531591,POINT (13.53159 52.56942),Neu-Hohenschönhausen,Lichtenberg,1109,OSM,...,NaN,https://11k14.info/,<NA>,<NA>,True,52.569419,13.531591,True,True,way
3,12797375598,11011011,15. Schule\nISS mit gymnasialer Oberstufe - 11K15,52.524963,13.513364,POINT (13.51336 52.52496),Lichtenberg,Lichtenberg,1103,OSM,...,sekretariat@11K15.schule.berlin.de,https://iss.adkosmos.eu/,<NA>,<NA>,True,52.524963,13.513364,True,True,node
4,1158766413,11001001,2. Schulpraktisches Seminar Mitte (S),52.546237,13.392748,POINT (13.39275 52.54624),Gesundbrunnen,Mitte,0106,OSM,...,NaN,NaN,<NA>,<NA>,True,52.546237,13.392748,True,True,way


In [23]:
# final dataset: only rows with district_id
schools_final = schools_unified_dedup[schools_unified_dedup["district_id"].notna()].copy()
print("Rows in unified_dedup:", len(schools_unified_dedup))
print("Rows in final (with district_id):", len(schools_final))

Rows in unified_dedup: 1032
Rows in final (with district_id): 1032


In [24]:
# drop unneeded columns
drop_cols = [
    "source",
    "source_layer",
    "primary_source",
    "source_ids",
    "source_type",
    "is_potential_duplicate",
    "has_district",
    "has_neighborhood",
]
# only drop columns that exist
drop_cols_existing = [c for c in drop_cols if c in schools_final.columns]
schools_final = schools_final.drop(columns=drop_cols_existing)
schools_final.head()

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,external_id,...,is_barrier_free,accessibility_notes,phone,email,website,last_source_update,last_ingested_at,is_active,lat,lon
0,12797375599,11011011,12. Gymnasium Lichtenberg 11Y12,52.524837,13.513904,POINT (13.5139 52.52484),Lichtenberg,Lichtenberg,1103,NaN,...,<NA>,<NA>,+49 30 2639114000,sekretariat@11y12.schule.berlin.de,https://gym.adkosmos.eu/,<NA>,<NA>,True,52.524837,13.513904
1,41406792,11011011,12. Schule mit sonderpädagogischem Förderschwe...,52.502780,13.509228,POINT (13.50923 52.50278),Friedrichsfelde,Lichtenberg,1101,NaN,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.502780,13.509228
2,656849860,11011011,14. Schule (Integrierte Sekundarschule),52.569419,13.531591,POINT (13.53159 52.56942),Neu-Hohenschönhausen,Lichtenberg,1109,11K14,...,<NA>,<NA>,NaN,NaN,https://11k14.info/,<NA>,<NA>,True,52.569419,13.531591
3,12797375598,11011011,15. Schule\nISS mit gymnasialer Oberstufe - 11K15,52.524963,13.513364,POINT (13.51336 52.52496),Lichtenberg,Lichtenberg,1103,NaN,...,<NA>,<NA>,+4930 - 263 976 00 00,sekretariat@11K15.schule.berlin.de,https://iss.adkosmos.eu/,<NA>,<NA>,True,52.524963,13.513364
4,1158766413,11001001,2. Schulpraktisches Seminar Mitte (S),52.546237,13.392748,POINT (13.39275 52.54624),Gesundbrunnen,Mitte,0106,NaN,...,<NA>,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.546237,13.392748


In [25]:
def _is_missing_scalar(v) -> bool:
    if v is None:
        return True
    if isinstance(v, float) and pd.isna(v):
        return True
    if isinstance(v, str) and v.strip() == "":
        return True
    return False

def _is_missing(series: pd.Series) -> pd.Series:
    return series.apply(_is_missing_scalar)

def reverse_enrich_address_postal(
    df: pd.DataFrame,
    lat_col: str = "latitude",
    lon_col: str = "longitude",
    addr_col: str = "address",
    pc_col: str = "postal_code",
    user_agent: str = "schools_reverse_enrich_max",
) -> pd.DataFrame:
    df = df.copy()

    need_addr = _is_missing(df[addr_col]) if addr_col in df.columns else pd.Series(True, index=df.index)
    need_pc   = _is_missing(df[pc_col]) if pc_col in df.columns else pd.Series(True, index=df.index)
    has_coords = df[lat_col].notna() & df[lon_col].notna()

    mask = (need_addr | need_pc) & has_coords
    if not mask.any():
        return df

    geolocator = Nominatim(user_agent=user_agent, timeout=15)
    reverse = RateLimiter(
        geolocator.reverse,
        min_delay_seconds=1.1,
        max_retries=4,
        error_wait_seconds=5.0,
        swallow_exceptions=True,
        return_value_on_exception=None,
    )

    coords = df.loc[mask, [lat_col, lon_col]].copy()
    coords["lat_r"] = coords[lat_col].astype(float).round(6)
    coords["lon_r"] = coords[lon_col].astype(float).round(6)
    uniq = coords.drop_duplicates(subset=["lat_r", "lon_r"])

    cache = {}

    def _rev(lat, lon):
        key = (lat, lon)
        if key in cache:
            return cache[key]

        loc = reverse((lat, lon), language="de", addressdetails=True, zoom=18)
        raw = getattr(loc, "raw", {}) if loc is not None else {}
        adr = (raw.get("address", {}) or {}) if isinstance(raw, dict) else {}

        road = adr.get("road")
        house = adr.get("house_number")
        postcode = adr.get("postcode")

        address_fill = pd.NA
        if road:
            address_fill = f"{road} {house}".strip() if house else str(road)

        cache[key] = (address_fill, postcode)
        return cache[key]

    fills = uniq.apply(
        lambda r: pd.Series(_rev(r["lat_r"], r["lon_r"]), index=["address_fill", "postal_code_fill"]),
        axis=1,
    )
    uniq = pd.concat([uniq.reset_index(drop=True), fills.reset_index(drop=True)], axis=1)

    coords = coords.merge(
        uniq[["lat_r", "lon_r", "address_fill", "postal_code_fill"]],
        on=["lat_r", "lon_r"],
        how="left",
    )

    if addr_col in df.columns:
        df.loc[mask, addr_col] = df.loc[mask, addr_col].where(
            ~_is_missing(df.loc[mask, addr_col]),
            coords["address_fill"].values
        )
    else:
        df[addr_col] = pd.NA
        df.loc[mask, addr_col] = coords["address_fill"].values

    if pc_col in df.columns:
        df.loc[mask, pc_col] = df.loc[mask, pc_col].where(
            ~_is_missing(df.loc[mask, pc_col]),
            coords["postal_code_fill"].values
        )
    else:
        df[pc_col] = pd.NA
        df.loc[mask, pc_col] = coords["postal_code_fill"].values

    return df

In [26]:
# reverse geocoding to fill missing addresses/postal codes
schools_final = reverse_enrich_address_postal(
    schools_final,
    lat_col="latitude",
    lon_col="longitude",
    addr_col="address",
    pc_col="postal_code",
    user_agent="schools_reverse_final_max",
)

print("missing_address:", int(_is_missing(schools_final["address"]).sum()))
print("missing_postal_code:", int(_is_missing(schools_final["postal_code"]).sum()))

missing_address: 0
missing_postal_code: 0


In [27]:
# Fill missing school names
schools_final["name"] = schools_final["name"].astype("string").str.strip()

missing_name_mask = schools_final["name"].isna() | (schools_final["name"] == "")
missing_name_count = int(missing_name_mask.sum())

schools_final.loc[missing_name_mask, "name"] = "Unknown_schools"

print("Filled missing school names with placeholder:", missing_name_count)
print("Remaining missing names:", int(schools_final["name"].isna().sum()))

Filled missing school names with placeholder: 20
Remaining missing names: 0


In [28]:
# Enrich missing data in schools_unified from schools_final
fill_cols = ["address", "postal_code", "phone", "email", "website", "opening_hours"]

assert "id" in schools_unified.columns, "schools_unified has no 'id' column"
assert "id" in schools_final.columns, "schools_final has no 'id' column"

cols_to_pull = [c for c in fill_cols if c in schools_unified.columns and c in schools_final.columns]
assert len(cols_to_pull) > 0, "No overlapping enrichment columns found between schools_unified and schools_final"

final_lookup = schools_final[["id"] + cols_to_pull].copy()
final_lookup["id"] = final_lookup["id"].astype("string").str.strip()
final_lookup = final_lookup.drop_duplicates(subset=["id"])

schools_unified_enriched = schools_unified.copy()
schools_unified_enriched["id"] = schools_unified_enriched["id"].astype("string").str.strip()

schools_unified_enriched = schools_unified_enriched.merge(
    final_lookup,
    on="id",
    how="left",
    suffixes=("", "_final"),
)

for c in cols_to_pull:
    base = schools_unified_enriched[c]
    incoming = schools_unified_enriched[f"{c}_final"]

    if base.dtype.name in ("string", "object"):
        base_norm = base.astype("string").fillna("").str.strip()
        missing_mask = base.isna() | (base_norm == "")
    else:
        missing_mask = base.isna()

    schools_unified_enriched.loc[missing_mask, c] = incoming.loc[missing_mask]

schools_unified_enriched = schools_unified_enriched.drop(columns=[f"{c}_final" for c in cols_to_pull])

# overwrite for downstream steps
schools_unified = schools_unified_enriched

# quick summary stats
summary = {
    "rows": len(schools_unified),
    "missing_address": (schools_unified["address"].astype("string").fillna("").str.strip() == "").sum() if "address" in schools_unified.columns else None,
    "missing_postal_code": (schools_unified["postal_code"].astype("string").fillna("").str.strip() == "").sum() if "postal_code" in schools_unified.columns else None,
}
pd.DataFrame([summary])

,rows,missing_address,missing_postal_code
0,1072,36,36


In [29]:
# helper functions
def _is_missing(s: pd.Series) -> pd.Series:
    return s.isna() | (s.astype("string").fillna("").str.strip() == "")

def _is_missing_scalar(v) -> bool:
    if pd.isna(v):
        return True
    return str(v).strip() == ""

In [30]:
def reverse_enrich_address_postal(
    df: pd.DataFrame,
    lat_col: str = "latitude",
    lon_col: str = "longitude",
    addr_col: str = "address",
    pc_col: str = "postal_code",
    user_agent: str = "schools_reverse_enrich_max",
) -> pd.DataFrame:
    df = df.copy()

    need_addr = _is_missing(df[addr_col]) if addr_col in df.columns else True
    need_pc   = _is_missing(df[pc_col]) if pc_col in df.columns else True
    has_coords = df[lat_col].notna() & df[lon_col].notna()

    mask = (need_addr | need_pc) & has_coords
    if not mask.any():
        return df

    geolocator = Nominatim(user_agent=user_agent, timeout=15)
    reverse = RateLimiter(
        geolocator.reverse,
        min_delay_seconds=1.1,
        max_retries=4,
        error_wait_seconds=5.0,
        swallow_exceptions=True,
        return_value_on_exception=None,
    )

    # Only geocode unique rounded coordinates
    coords = df.loc[mask, [lat_col, lon_col]].copy()
    coords["lat_r"] = coords[lat_col].astype(float).round(6)
    coords["lon_r"] = coords[lon_col].astype(float).round(6)
    uniq = coords.drop_duplicates(subset=["lat_r", "lon_r"])

    cache = {}

    def _rev(lat, lon):
        key = (lat, lon)
        if key in cache:
            return cache[key]

        loc = reverse((lat, lon), language="de", addressdetails=True, zoom=18)
        raw = getattr(loc, "raw", {}) if loc is not None else {}
        adr = (raw.get("address", {}) or {}) if isinstance(raw, dict) else {}

        road = adr.get("road")
        house = adr.get("house_number")
        postcode = adr.get("postcode")

        address_fill = pd.NA
        if road:
            address_fill = f"{road} {house}".strip() if house else str(road)

        cache[key] = (address_fill, postcode)
        return cache[key]

    fills = uniq.apply(
        lambda r: pd.Series(_rev(r["lat_r"], r["lon_r"]), index=["address_fill", "postal_code_fill"]),
        axis=1,
    )
    uniq = pd.concat([uniq.reset_index(drop=True), fills.reset_index(drop=True)], axis=1)

    # Map back to all masked rows via rounded coords
    coords = coords.merge(uniq[["lat_r", "lon_r", "address_fill", "postal_code_fill"]], on=["lat_r", "lon_r"], how="left")

    df.loc[mask, addr_col] = df.loc[mask, addr_col].where(~_is_missing(df.loc[mask, addr_col]), coords["address_fill"].values)
    df.loc[mask, pc_col]   = df.loc[mask, pc_col].where(~_is_missing(df.loc[mask, pc_col]), coords["postal_code_fill"].values)

    return df

In [31]:
# derive postal_code from address when postal_code is missing
import re

pc_missing = _is_missing(schools_unified["postal_code"])
addr_has_value = ~_is_missing(schools_unified["address"])

extracted = schools_unified.loc[pc_missing & addr_has_value, "address"].astype("string").str.extract(r"(\b\d{5}\b)")
schools_unified.loc[pc_missing & addr_has_value, "postal_code"] = extracted[0]

pd.DataFrame([{
    "missing_postal_code_after_extract": int(_is_missing(schools_unified["postal_code"]).sum())
}])

,missing_postal_code_after_extract
0,36


In [32]:
# reverse geocode again to fill any remaining missing addresses/postal codes
schools_unified = reverse_enrich_address_postal(
    schools_unified,
    lat_col="latitude",
    lon_col="longitude",
    addr_col="address",
    pc_col="postal_code",
    user_agent="schools_reverse_unified_max",
)

print("missing_address:", int(_is_missing(schools_unified["address"]).sum()))
print("missing_postal_code:", int(_is_missing(schools_unified["postal_code"]).sum()))

missing_address: 0
missing_postal_code: 0


In [33]:
# join check for missing names, addresses, postal codes
tmp = schools_unified.copy()

tmp["is_unknown_name"] = (
    tmp["name"].astype("string").fillna("").str.strip().str.lower().eq("unknown_school")
    )
tmp["missing_address"] = _is_missing(tmp["address"]) if "address" in tmp.columns else True
tmp["missing_postal_code"] = _is_missing(tmp["postal_code"]) if "postal_code" in tmp.columns else True

summary = pd.DataFrame([{
    "rows": len(tmp),
    "unknown_name_rows": int(tmp["is_unknown_name"].sum()),
    "missing_address_total": int(tmp["missing_address"].sum()),
    "missing_address_unknown": int((tmp["missing_address"] & tmp["is_unknown_name"]).sum()),
    "missing_address_not_unknown": int((tmp["missing_address"] & ~tmp["is_unknown_name"]).sum()),
    "missing_postal_code_total": int(tmp["missing_postal_code"].sum()),
    "missing_postal_code_unknown": int((tmp["missing_postal_code"] & tmp["is_unknown_name"]).sum()),
    "missing_postal_code_not_unknown": int((tmp["missing_postal_code"] & ~tmp["is_unknown_name"]).sum()),
}])

summary

,rows,unknown_name_rows,missing_address_total,missing_address_unknown,missing_address_not_unknown,missing_postal_code_total,missing_postal_code_unknown,missing_postal_code_not_unknown
0,1072,0,0,0,0,0,0,0


In [34]:
# Enrich missing data in schools_final from schools_unified
fill_cols = ["address", "postal_code"]

su_fill = schools_unified[["id"] + fill_cols].copy()
su_fill["id"] = su_fill["id"].astype("string").str.strip()
su_fill = su_fill.drop_duplicates(subset=["id"])

sf = schools_final.copy()
sf["id"] = sf["id"].astype("string").str.strip()

sf = sf.merge(su_fill, on="id", how="left", suffixes=("", "_unified"))

for c in fill_cols:
    if c in sf.columns:
        m = _is_missing(sf[c])
        sf.loc[m, c] = sf.loc[m, f"{c}_unified"]
    else:
        sf[c] = sf[f"{c}_unified"]
    sf = sf.drop(columns=[f"{c}_unified"])

schools_final = sf

print("schools_final missing_address:", int(_is_missing(schools_final["address"]).sum()))
print("schools_final missing_postal_code:", int(_is_missing(schools_final["postal_code"]).sum()))

schools_final missing_address: 0
schools_final missing_postal_code: 0


## Summary

- OSM-based schools were cleaned, mapped to the common POI schema, and spatially joined with LOR polygons to enrich each record with `district`, `district_id`, `neighborhood`, and `neighborhood_id`.
- Initially, some schools did not receive a `district_id` despite being located within the Berlin boundary. The spatial join was refined to use a centroid-based approach for polygon geometries (while keeping point geometries unchanged). After this adjustment, all schools have a valid `district_id` (`schools_unified["district_id"].isna().sum() == 0`), so no records are excluded from the final export.
- Potential duplicates were identified using `(name, address, district)` and consolidated into a deduplicated table (`schools_unified_dedup`), keeping one record per group and preferring more informative OSM features (e.g., entries with richer contact attributes).
- The final table `schools_final`:
  - contains one row per school (after deduplication),
  - guarantees a valid `district_id` for every record,
  - drops transformation- and source-specific helper columns,
  - retains the POI core fields (id, coordinates, district/neighborhood attributes) and key school attributes (address, ownership/operator, and contact fields where available).
- Missing postal codes were partially recovered by extracting 5-digit codes from existing address strings where possible, and remaining gaps were addressed via reverse geocoding.
- Reverse geocoding (Nominatim) was used to enrich missing address information from latitude/longitude and to fill remaining missing postal codes where feasible, using rate limiting (delays + retry logic) to respect the service usage policy.
- Missing school names were filled with the placeholder `Unknown` to avoid NULLs in the final export while keeping all records.

In [35]:
# Save final dataset
df_upload = schools_final.copy()
output_path_final = "data/schools_final_osm_only.csv"
schools_final.to_csv(output_path_final, index=False)
print(f"Saved dataset with {len(schools_final)} rows to {output_path_final}")

Saved dataset with 1032 rows to data/schools_final_osm_only.csv


In [36]:
import os
from getpass import getpass

import pandas as pd
from sqlalchemy import create_engine, text
from sqlalchemy.engine import URL


In [38]:
TARGET_SCHEMA = "berlin_source_data"

TARGET_TABLE  = f"schools_kai"

In [ ]:
engine = create_engine('postgresql+psycopg2://user:password@localhost:5433/layereddb')

In [44]:
# upload dataframe
df_upload = schools_final.copy()

if "geometry" in df_upload.columns:
    df_upload["geometry"] = df_upload["geometry"].astype(str)
    df_upload = df_upload.drop(columns=["geometry"])

if "latitude" in df_upload.columns and "longitude" in df_upload.columns:
    df_upload["coordinates"] = df_upload["latitude"].astype(str) + "," + df_upload["longitude"].astype(str)

if "id" in df_upload.columns:
    df_upload["id"] = df_upload["id"].astype(str)

df_upload.head()

/var/folders/tk/bx1zbkk52y3b84pwkrt4dz5c0000gn/T/ipykernel_22880/1044379625.py:5: UserWarning: Geometry column does not contain geometry.
  df_upload["geometry"] = df_upload["geometry"].astype(str)


,id,district_id,name,latitude,longitude,neighborhood,district,neighborhood_id,external_id,school_name,...,accessibility_notes,phone,email,website,last_source_update,last_ingested_at,is_active,lat,lon,coordinates
0,12797375599,11011011,12. Gymnasium Lichtenberg 11Y12,52.524837,13.513904,Lichtenberg,Lichtenberg,1103,NaN,12. Gymnasium Lichtenberg 11Y12,...,<NA>,+49 30 2639114000,sekretariat@11y12.schule.berlin.de,https://gym.adkosmos.eu/,<NA>,<NA>,True,52.524837,13.513904,"52.5248369,13.513904399999998"
1,41406792,11011011,12. Schule mit sonderpädagogischem Förderschwe...,52.502780,13.509228,Friedrichsfelde,Lichtenberg,1101,NaN,12. Schule mit sonderpädagogischem Förderschwe...,...,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.502780,13.509228,"52.50278008691506,13.509228123599339"
2,656849860,11011011,14. Schule (Integrierte Sekundarschule),52.569419,13.531591,Neu-Hohenschönhausen,Lichtenberg,1109,11K14,14. Schule (Integrierte Sekundarschule),...,<NA>,NaN,NaN,https://11k14.info/,<NA>,<NA>,True,52.569419,13.531591,"52.56941881673431,13.531591483715927"
3,12797375598,11011011,15. Schule\nISS mit gymnasialer Oberstufe - 11K15,52.524963,13.513364,Lichtenberg,Lichtenberg,1103,NaN,15. Schule\nISS mit gymnasialer Oberstufe - 11K15,...,<NA>,+4930 - 263 976 00 00,sekretariat@11K15.schule.berlin.de,https://iss.adkosmos.eu/,<NA>,<NA>,True,52.524963,13.513364,"52.524962899999984,13.5133643"
4,1158766413,11001001,2. Schulpraktisches Seminar Mitte (S),52.546237,13.392748,Gesundbrunnen,Mitte,0106,NaN,2. Schulpraktisches Seminar Mitte (S),...,<NA>,NaN,NaN,NaN,<NA>,<NA>,True,52.546237,13.392748,"52.5462373327245,13.392747812327087"


In [45]:
df_upload.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1032 entries, 0 to 1031
Data columns (total 37 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   1032 non-null   object 
 1   district_id          1032 non-null   string 
 2   name                 1032 non-null   string 
 3   latitude             1032 non-null   float64
 4   longitude            1032 non-null   float64
 5   neighborhood         1032 non-null   object 
 6   district             1032 non-null   object 
 7   neighborhood_id      1032 non-null   object 
 8   external_id          867 non-null    object 
 9   school_name          1012 non-null   object 
 10  school_type          0 non-null      object 
 11  school_subtype       0 non-null      object 
 12  grades_offered       0 non-null      object 
 13  is_primary           0 non-null      object 
 14  is_secondary         0 non-null      object 
 15  is_vocational        0 non-null      o

In [46]:
df_upload = df_upload.drop(columns=['lat', 'lon'])

In [47]:
df_upload.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1032 entries, 0 to 1031
Data columns (total 35 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   1032 non-null   object 
 1   district_id          1032 non-null   string 
 2   name                 1032 non-null   string 
 3   latitude             1032 non-null   float64
 4   longitude            1032 non-null   float64
 5   neighborhood         1032 non-null   object 
 6   district             1032 non-null   object 
 7   neighborhood_id      1032 non-null   object 
 8   external_id          867 non-null    object 
 9   school_name          1012 non-null   object 
 10  school_type          0 non-null      object 
 11  school_subtype       0 non-null      object 
 12  grades_offered       0 non-null      object 
 13  is_primary           0 non-null      object 
 14  is_secondary         0 non-null      object 
 15  is_vocational        0 non-null      o

In [58]:
# Validate schema and table name
assert TARGET_SCHEMA == "berlin_source_data", "TARGET_SCHEMA is not berlin_source_data"
assert all(c.isalnum() or c == "_" for c in TARGET_TABLE), "Invalid table name"

create_table_sql = f"""
CREATE TABLE IF NOT EXISTS "{TARGET_SCHEMA}"."{TARGET_TABLE}" (
    id                 VARCHAR(64) PRIMARY KEY,
    district_id         VARCHAR(10) NOT NULL
        REFERENCES "{TARGET_SCHEMA}".districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE,

    name               VARCHAR(255) NOT NULL,

    school_name         VARCHAR(255),
    school_type         VARCHAR(255),
    school_subtype      VARCHAR(255),
    operator            VARCHAR(255),
    ownership           VARCHAR(255),
    grades_offered      VARCHAR,       

    address             VARCHAR(255),
    street              VARCHAR(200),
    house_number        VARCHAR(50),
    postal_code         VARCHAR(10),
    city                VARCHAR(100),

    phone               VARCHAR(50),
    email               VARCHAR(255),
    website             VARCHAR(255),
    opening_hours       VARCHAR(255),

    source              VARCHAR(50),
    source_layer        VARCHAR(100),
    external_id         VARCHAR(100),
    primary_source      VARCHAR(50),
    source_ids          TEXT,

    latitude            DECIMAL(9,6),
    longitude           DECIMAL(9,6),

    geometry            TEXT,

    neighborhood        VARCHAR(120),
    district            VARCHAR(120),
    neighborhood_id     VARCHAR(50),

    official_school_id  VARCHAR(50),
    is_barrier_free     BOOLEAN,
    accessibility_notes TEXT,

    last_source_update  TIMESTAMP,
    last_ingested_at    TIMESTAMP,
    is_active           BOOLEAN
    
);
"""
# Create the target table
with engine.begin() as conn:
    conn.execute(text(create_table_sql))

print(f"Table ensured: {TARGET_SCHEMA}.{TARGET_TABLE}")

Table ensured: berlin_source_data.schools_kai


In [59]:
# table columns
with engine.connect() as conn:
    table_cols = pd.read_sql(
        text("""
            SELECT column_name
            FROM information_schema.columns
            WHERE table_schema = :schema
              AND table_name   = :table
            ORDER BY ordinal_position
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )["column_name"].tolist()

print("DB columns:", table_cols)

# dataframe to upload
df_upload = schools_final.copy()

if "geometry" in df_upload.columns:
    df_upload["geometry"] = df_upload["geometry"].astype(str)

# required columns: id, name, district_id
if "id" not in df_upload.columns:
    for candidate in ["school_id", "osm_id", "external_id"]:
        if candidate in df_upload.columns:
            df_upload["id"] = df_upload[candidate].astype(str)
            break
if "id" not in df_upload.columns:
    raise ValueError("no 'id' in DataFrame finded.")

df_upload["id"] = df_upload["id"].astype(str)

# name and district_id
if "name" not in df_upload.columns:
    if "school_name" in df_upload.columns:
        df_upload["name"] = df_upload["school_name"]
    else:
        df_upload["name"] = None
df_upload["name"] = df_upload["name"].fillna("Unknown").astype(str)
if "district_id" not in df_upload.columns:
    raise ValueError("column 'district_id' is missing in DataFrame.")
df_upload["district_id"] = df_upload["district_id"].astype(str)
if "coordinates" in table_cols and "coordinates" not in df_upload.columns:
    if {"latitude", "longitude"}.issubset(df_upload.columns):
        df_upload["coordinates"] = df_upload["latitude"].astype(str) + "," + df_upload["longitude"].astype(str)
    elif {"lat", "lon"}.issubset(df_upload.columns):
        df_upload["coordinates"] = df_upload["lat"].astype(str) + "," + df_upload["lon"].astype(str)

# add missing columns with NULLs
for c in table_cols:
    if c not in df_upload.columns:
        df_upload[c] = None

df_upload = df_upload[table_cols]

# sanity checks before insert
assert df_upload["id"].notna().all(), "id has NULL"
assert df_upload["district_id"].notna().all(), "district_id has NULL"
assert df_upload["name"].notna().all(), "name has NULL"
assert not df_upload["id"].duplicated().any(), "id has Dup"

print("df_upload ready")
print("rows:", len(df_upload), "cols:", df_upload.shape[1])
df_upload.head(3)

DB columns: ['id', 'district_id', 'name', 'school_name', 'school_type', 'school_subtype', 'operator', 'ownership', 'grades_offered', 'address', 'street', 'house_number', 'postal_code', 'city', 'phone', 'email', 'website', 'opening_hours', 'source', 'source_layer', 'external_id', 'primary_source', 'source_ids', 'latitude', 'longitude', 'geometry', 'neighborhood', 'district', 'neighborhood_id', 'official_school_id', 'is_barrier_free', 'accessibility_notes', 'last_source_update', 'last_ingested_at', 'is_active']
df_upload ready
rows: 1032 cols: 35


/var/folders/tk/bx1zbkk52y3b84pwkrt4dz5c0000gn/T/ipykernel_22880/384410406.py:21: UserWarning: Geometry column does not contain geometry.
  df_upload["geometry"] = df_upload["geometry"].astype(str)


,id,district_id,name,school_name,school_type,school_subtype,operator,ownership,grades_offered,address,...,geometry,neighborhood,district,neighborhood_id,official_school_id,is_barrier_free,accessibility_notes,last_source_update,last_ingested_at,is_active
0,12797375599,11011011,12. Gymnasium Lichtenberg 11Y12,12. Gymnasium Lichtenberg 11Y12,<NA>,<NA>,None,Bezirksamt Lichtenberg,<NA>,"Allee der Kosmonauten 22, 10315 Berlin",...,POINT (13.513904 52.524837),Lichtenberg,Lichtenberg,1103,<NA>,<NA>,<NA>,<NA>,<NA>,True
1,41406792,11011011,12. Schule mit sonderpädagogischem Förderschwe...,12. Schule mit sonderpädagogischem Förderschwe...,<NA>,<NA>,None,NaN,<NA>,Rummelsburger Straße 21,...,POINT (13.509228 52.50278),Friedrichsfelde,Lichtenberg,1101,<NA>,<NA>,<NA>,<NA>,<NA>,True
2,656849860,11011011,14. Schule (Integrierte Sekundarschule),14. Schule (Integrierte Sekundarschule),<NA>,<NA>,None,NaN,<NA>,"Wartiner Straße 1-3, 13057 Berlin",...,POINT (13.531591 52.569419),Neu-Hohenschönhausen,Lichtenberg,1109,<NA>,<NA>,<NA>,<NA>,<NA>,True


In [60]:
with engine.connect() as conn:
    col_limits = pd.read_sql(
        text("""
            SELECT column_name, data_type, character_maximum_length
            FROM information_schema.columns
            WHERE table_schema = :schema
              AND table_name   = :table
              AND data_type IN ('character varying', 'character')
              AND character_maximum_length IS NOT NULL
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )

# nur varchar(100)
v100 = col_limits[col_limits["character_maximum_length"] == 100]["column_name"].tolist()

offenders = []
for c in v100:
    if c in df_upload.columns:
        s = df_upload[c].astype("string")
        max_len = s.str.len().dropna().max()
        if pd.notna(max_len) and int(max_len) > 100:
            offenders.append((c, int(max_len)))

offenders_df = pd.DataFrame(offenders, columns=["column", "max_len_in_df"]).sort_values("max_len_in_df", ascending=False)
offenders_df


,column,max_len_in_df


In [61]:
with engine.begin() as conn:
    conn.execute(text(f'''
        ALTER TABLE "{TARGET_SCHEMA}"."{TARGET_TABLE}"
        ALTER COLUMN "ownership" TYPE VARCHAR(255);
    '''))


In [63]:
df_upload.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1032 entries, 0 to 1031
Data columns (total 35 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   1032 non-null   object 
 1   district_id          1032 non-null   object 
 2   name                 1032 non-null   object 
 3   school_name          1012 non-null   object 
 4   school_type          0 non-null      object 
 5   school_subtype       0 non-null      object 
 6   operator             0 non-null      object 
 7   ownership            326 non-null    object 
 8   grades_offered       0 non-null      object 
 9   address              1032 non-null   object 
 10  street               834 non-null    object 
 11  house_number         831 non-null    object 
 12  postal_code          1032 non-null   object 
 13  city                 806 non-null    object 
 14  phone                338 non-null    object 
 15  email                266 non-null    o

In [62]:
# insert data into target table
from sqlalchemy import text

with engine.begin() as conn:
    existing = conn.execute(
        text(f'SELECT COUNT(*) FROM "{TARGET_SCHEMA}"."{TARGET_TABLE}"')
    ).scalar()
    print("Existing rows in target table:", existing)
    conn.execute(text(f'TRUNCATE TABLE "{TARGET_SCHEMA}"."{TARGET_TABLE}"'))

# insert data
df_upload.to_sql(
    name=TARGET_TABLE,
    con=engine,
    schema=TARGET_SCHEMA,
    if_exists="append",
    index=False,
    method="multi",
    chunksize=2000,
)

# verify row count after insert
with engine.connect() as conn:
    new_count = conn.execute(
        text(f'SELECT COUNT(*) FROM "{TARGET_SCHEMA}"."{TARGET_TABLE}"')
    ).scalar()
    print("Insert done. Rows now in target table:", new_count)

Existing rows in target table: 0
Insert done. Rows now in target table: 1032


In [64]:
# Quick data quality assessment query
with engine.connect() as conn:
    qa = pd.read_sql(text(f"""
        WITH base AS (
            SELECT *
            FROM "{TARGET_SCHEMA}"."{TARGET_TABLE}"
        ),
        dupes AS (
            SELECT COUNT(*) AS dupe_id_rows
            FROM (
                SELECT id
                FROM base
                GROUP BY id
                HAVING COUNT(*) > 1
            ) x
        ),
        fk_district AS (
            SELECT COUNT(*) AS missing_district_fk
            FROM base s
            LEFT JOIN berlin_source_data.districts d
              ON d.district_id = s.district_id
            WHERE s.district_id IS NOT NULL
              AND d.district_id IS NULL
        ),
        fk_neighborhood AS (
            SELECT COUNT(*) AS missing_neighborhood_fk
            FROM base s
            LEFT JOIN berlin_source_data.neighborhoods n
              ON n.neighborhood_id = s.neighborhood_id
            WHERE s.neighborhood_id IS NOT NULL
              AND n.neighborhood_id IS NULL
        )
        SELECT
            (SELECT COUNT(*) FROM base) AS total_rows,
            (SELECT dupe_id_rows FROM dupes) AS dupe_id_rows,

            COUNT(*) FILTER (WHERE name IS NULL OR name = '') AS null_name,
            COUNT(*) FILTER (WHERE name = 'Unknown') AS unknown_name,

            COUNT(*) FILTER (WHERE latitude IS NULL) AS null_lat,
            COUNT(*) FILTER (WHERE longitude IS NULL) AS null_lon,
            COUNT(*) FILTER (WHERE latitude  NOT BETWEEN 52.3 AND 52.7) AS lat_out_of_range,
            COUNT(*) FILTER (WHERE longitude NOT BETWEEN 13.0 AND 13.8) AS lon_out_of_range,

            COUNT(*) FILTER (WHERE district_id IS NULL) AS null_district_id,
            COUNT(*) FILTER (WHERE neighborhood_id IS NULL) AS null_neighborhood_id,

            COUNT(*) FILTER (WHERE address IS NULL OR address = '') AS missing_address,
            COUNT(*) FILTER (WHERE postal_code IS NULL OR postal_code = '') AS missing_postal_code,

            (SELECT missing_district_fk FROM fk_district) AS missing_district_fk,
            (SELECT missing_neighborhood_fk FROM fk_neighborhood) AS missing_neighborhood_fk
        FROM base;
    """), conn)

qa

,total_rows,dupe_id_rows,null_name,unknown_name,null_lat,null_lon,lat_out_of_range,lon_out_of_range,null_district_id,null_neighborhood_id,missing_address,missing_postal_code,missing_district_fk,missing_neighborhood_fk
0,1032,0,0,0,0,0,0,0,0,0,0,0,0,0


## Test Insert – Ergebnis (berlin_source_data.schools_<DB_USER>)

**Insert**
- Rows inserted: 1032

**Integrity / Schema**
- Duplicate IDs: 0
- `name` NULLs: 0
- Latitude/Longitude NULLs: 0
- Coordinates out of Berlin range: 0
- `district_id` NULLs: 0
- `neighborhood_id` NULLs: 0
- Missing FK in `districts`: 0
- Missing FK in `neighborhoods`: 0

**Known data quality gaps (non-blocking)**
- `name = 'Unknown'`: 20
- Missing `address`: 0
- Missing `postal_code`: 0

**Conclusion**
Table creation + constraints + FK integrity validated successfully. Remaining gaps are content-level fields (address/postal_code/name) and can be enriched later without breaking relational consistency.

In [65]:
# Get table schema, constraints, and indexes
with engine.connect() as conn:
    cols = pd.read_sql(
        text("""
            SELECT
                ordinal_position,
                column_name,
                data_type,
                is_nullable,
                character_maximum_length,
                numeric_precision,
                numeric_scale
            FROM information_schema.columns
            WHERE table_schema = :schema
              AND table_name   = :table
            ORDER BY ordinal_position;
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )
    constraints = pd.read_sql(
        text("""
            SELECT
                c.conname AS constraint_name,
                c.contype AS constraint_type,
                pg_get_constraintdef(c.oid, true) AS definition
            FROM pg_constraint c
            JOIN pg_class t      ON t.oid = c.conrelid
            JOIN pg_namespace ns ON ns.oid = t.relnamespace
            WHERE ns.nspname = :schema
              AND t.relname  = :table
            ORDER BY c.contype, c.conname;
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )

    indexes = pd.read_sql(
        text("""
            SELECT indexname, indexdef
            FROM pg_indexes
            WHERE schemaname = :schema
              AND tablename  = :table
            ORDER BY indexname;
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )

cols, constraints, indexes

(    ordinal_position          column_name                    data_type  \
 0                  1                   id            character varying   
 1                  2          district_id            character varying   
 2                  3                 name            character varying   
 3                  4          school_name            character varying   
 4                  5          school_type            character varying   
 5                  6       school_subtype            character varying   
 6                  7             operator            character varying   
 7                  8            ownership            character varying   
 8                  9       grades_offered            character varying   
 9                 10              address            character varying   
 10                11               street            character varying   
 11                12         house_number            character varying   
 12                13    

In [66]:
# Add check constraints for latitude and longitude ranges if not exist
def constraint_exists(conn, schema: str, table: str, cname: str) -> bool:
    return conn.execute(
        text("""
            SELECT EXISTS (
                SELECT 1
                FROM pg_constraint c
                JOIN pg_class t      ON t.oid = c.conrelid
                JOIN pg_namespace ns ON ns.oid = t.relnamespace
                WHERE ns.nspname = :schema
                  AND t.relname  = :table
                  AND c.conname  = :cname
            );
        """),
        {"schema": schema, "table": table, "cname": cname},
    ).scalar()

chk_lat = f"{TARGET_TABLE}_chk_lat_range"
chk_lon = f"{TARGET_TABLE}_chk_lon_range"

# add constraints if not exist
with engine.begin() as conn:
    if not constraint_exists(conn, TARGET_SCHEMA, TARGET_TABLE, chk_lat):
        conn.execute(text(f"""
            ALTER TABLE "{TARGET_SCHEMA}"."{TARGET_TABLE}"
            ADD CONSTRAINT "{chk_lat}"
            CHECK (latitude IS NULL OR latitude BETWEEN 52.3 AND 52.7);
        """))
        print(f"Added {chk_lat}")
    else:
        print(f"Exists {chk_lat}")

    if not constraint_exists(conn, TARGET_SCHEMA, TARGET_TABLE, chk_lon):
        conn.execute(text(f"""
            ALTER TABLE "{TARGET_SCHEMA}"."{TARGET_TABLE}"
            ADD CONSTRAINT "{chk_lon}"
            CHECK (longitude IS NULL OR longitude BETWEEN 13.0 AND 13.8);
        """))
        print(f"Added {chk_lon}")
    else:
        print(f"Exists {chk_lon}")

Added schools_kai_chk_lat_range
Added schools_kai_chk_lon_range


In [67]:
pd.set_option("display.max_colwidth", None)

with engine.connect() as conn:
    constraints = pd.read_sql(
        text("""
            SELECT
                c.conname AS constraint_name,
                c.contype AS constraint_type,
                pg_get_constraintdef(c.oid, true) AS definition
            FROM pg_constraint c
            JOIN pg_class t      ON t.oid = c.conrelid
            JOIN pg_namespace ns ON ns.oid = t.relnamespace
            WHERE ns.nspname = :schema
              AND t.relname  = :table
            ORDER BY c.contype, c.conname;
        """),
        conn,
        params={"schema": TARGET_SCHEMA, "table": TARGET_TABLE},
    )

print(constraints.to_string(index=False))

             constraint_name constraint_type                                                                                                          definition
   schools_kai_chk_lat_range               c                                                   CHECK (latitude IS NULL OR latitude >= 52.3 AND latitude <= 52.7)
   schools_kai_chk_lon_range               c                                                CHECK (longitude IS NULL OR longitude >= 13.0 AND longitude <= 13.8)
schools_kai_district_id_fkey               f FOREIGN KEY (district_id) REFERENCES berlin_source_data.districts(district_id) ON UPDATE CASCADE ON DELETE RESTRICT
            schools_kai_pkey               p                                                                                                    PRIMARY KEY (id)


## Database Integration (Schools)

### Target table
- Schema/Table: `berlin_source_data.schools_<DB_USER>`
- Rows inserted: `1032`

### Constraints (enforced in DB)
- Primary key: `PRIMARY KEY (id)`
- Foreign key:
  - `district_id` references `berlin_source_data.districts(district_id)`
  - `ON UPDATE CASCADE ON DELETE RESTRICT`
- Coordinate quality checks:
  - `CHECK (latitude IS NULL OR latitude BETWEEN 52.3 AND 52.7)`
  - `CHECK (longitude IS NULL OR longitude BETWEEN 13.0 AND 13.8)`

### Validation results
- Duplicate IDs: `0`
- NULLs:
  - `name`: `0`
  - `district_id`: `0`
  - `neighborhood_id`: `0`
  - `latitude`: `0`
  - `longitude`: `0`
- Coordinates out of Berlin range: `0`
- FK integrity:
  - Missing `district_id` references: `0`
  - Missing `neighborhood_id` references (validated via join check): `0`
 
### Known data quality gaps (non-blocking)
- `name = 'Unknown'`: `20`
- Missing `address`: `0`
- Missing `postal_code`: `0`